File: pb/auditlog/auditlog.go
package auditlog

import (
	"log"
	"os"
	"strings"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"golang.org/x/exp/slices"
)

// collection names to be audit logged
var collections = strings.Split(os.Getenv("AUDITLOG"), ",")

func Register(app *pocketbase.PocketBase) {
	app.OnServe().BindFunc(func(se *core.ServeEvent) error {
		app.OnRecordCreateRequest().BindFunc(func(e *core.RecordRequestEvent) error {
			if err := e.Next(); err != nil {
				return err
			}
			return doAudit(app, "insert", e)
		})
		app.OnRecordUpdateRequest().BindFunc(func(e *core.RecordRequestEvent) error {
			if err := e.Next(); err != nil {
				return err
			}
			return doAudit(app, "update", e)
		})
		app.OnRecordDeleteRequest().BindFunc(func(e *core.RecordRequestEvent) error {
			if err := e.Next(); err != nil {
				return err
			}
			return doAudit(app, "delete", e)
		})
		return se.Next()
	})
}

func diff(val1, val2 any) bool {
	// handle comparison of non-comparable types
	// TODO: add more cases to the switch below as we discover them
	switch v1 := val1.(type) {
	case []string:
		v2 := val2.([]string)
		if len(v1) != len(v2) {
			return true
		}
		for i, v1 := range v1 {
			if diff(v1, v2[i]) {
				return true
			}
		}
		// no diff, if reached the end of the loop
		return false
	case []interface{}:
		return false
	default:
		// use builtin comparison by default
		return val1 != val2
	}
}

func doAudit(app *pocketbase.PocketBase, event string, request *core.RecordRequestEvent) error {
	record := request.Record
	auth := request.Auth
	collection := record.Collection().Name
	// exclude logging "auditlog" and include only what's in AUDITLOG env var
	if collection != "auditlog" && slices.Contains(collections, collection) {
		var user, admin string
		if auth.IsSuperuser() {
			admin = auth.Id
		} else {
			user = auth.Id
		}
		log.Printf("AuditLog:%s:%s:%s:%s:%s\n", collection, record.Id, event, user, admin)
		target, err := app.FindCollectionByNameOrId("auditlog")
		if err != nil {
			return err
		}
		auditlog := core.NewRecord(target)
		auditlog.Set("collection", collection)
		auditlog.Set("record", record.Id)
		auditlog.Set("event", event)
		auditlog.Set("user", user)
		auditlog.Set("admin", admin)
		// detect changes
		original := record.Original().PublicExport()
		recordExport := record.PublicExport()
		for k, v := range original {
			if !diff(v, recordExport[k]) { // unmodified, then remove
				delete(original, k)
			}
		}
		auditlog.Set("data", recordExport)
		auditlog.Set("original", original)

		return app.Save(auditlog)
	}
	return nil
}

File: pb/pb_hooks/auditlog.js
/// <reference path="../pb_data/types.d.ts" />

const collections = $os.getenv("AUDITLOG")?.split(",") ?? [];

/**
 * Inserts an auditlog record for the given event (insert/update/delete)
 * that happens on the given record.
 *
 * @param {string} event
 * @param {models.Record} record
 * @param {echo.Context} c
 */
function doAudit(event, record, c) {
  // list of collections that are to be audit-logged
  const collection = record.collection().name;
  // exclude logging "auditlog" and include only what's in AUDITLOG env var
  if (collection != "auditlog" && collections.includes(collection)) {
    /** @type {models.Admin} */
    const admin = c.get("admin");
    /** @type {models.Record} */
    const user = c.get("authRecord");
    console.log("AuditLog", collection, record.id, event, user, admin);
    const auditlog = new Record(
      $app.dao().findCollectionByNameOrId("auditlog")
    );
    auditlog.set("collection", collection);
    auditlog.set("record", record.id);
    auditlog.set("event", event);
    auditlog.set("user", user?.id);
    auditlog.set("admin", admin?.id);
    // detect changes
    const original = record.originalCopy().publicExport();
    const recordExport = record.publicExport();
    for (const [k, v] of Object.entries(original)) {
      if (v == recordExport[k]) {
        // unchanged, then remove from "original"
        delete original[k];
      }
    }
    auditlog.set("data", recordExport);
    auditlog.set("original", original);
    $app.dao().save(auditlog);
  }
}

module.exports = {
  doAudit,
};
